src/p_range.jl

```diff
@@ -622,42 +620,35 @@ function block_with_constant_size(rank,np,n,ghost,periodic=map(i->false,ghost))
     p = CartesianIndices(np)[rank]
     own_ranges = map(local_range,Tuple(p),np,n)
     local_ranges = map(local_range,Tuple(p),np,n,ghost,periodic)
-    owners = map(Tuple(p),own_ranges,local_ranges) do p,or,lr
+    owners = map(Tuple(p), np, n, local_ranges) do p, np, n, lr
         myowners = zeros(Int32,length(lr))
-        for i in 1:length(lr)
-            if lr[i] in or
+        i = 1
+        for p in Iterators.cycle(1:np)
+            plr = local_range(p, np, n)
+            while mod(lr[i]-1, n)+1 in plr
                 myowners[i] = p
+                (i += 1) > length(myowners) && return myowners
             end
         end
-        if myowners[1] == 0
-            myowners[1] = p-1
-        end
-        if myowners[end] == 0
-            myowners[end] = p+1
-        end
-        myowners
-    end
-    n_ghost = 0
-    cis = CartesianIndices(map(length,local_ranges))
-    predicate(p,i,owners) = owners[i] == p
-    for ci in cis
-        flags = map(predicate,Tuple(p),Tuple(ci),owners)
-        if !all(flags)
-            n_ghost += 1
-        end
     end
+    n_local = prod(map(length, local_ranges))
+    n_own = prod(map(length, own_ranges))
+    n_ghost = n_local - n_own
+
     ghost_to_global = zeros(Int,n_ghost)
     ghost_to_owner = zeros(Int32,n_ghost)
-    n_local = prod(map(length,local_ranges))
     perm = zeros(Int32,n_local)
     i_ghost = 0
     i_own = 0
-    n_own = prod(map(length,own_ranges))
+
+    cis = CartesianIndices(map(length,local_ranges))
     lis = CircularArray(LinearIndices(n))
     local_cis = CartesianIndices(local_ranges)
-    owner_lis = CircularArray(LinearIndices(np))
```

Comment:
I think this was the only place where we used CircularArrays (please confirm). If so, can you remove it as a dependence?

Reply:
There is still another place where CircularArrays are used (two lines above actually).